source/agora/cli/checkvtable/main.d

```diff
@@ -0,0 +1,222 @@
+/*******************************************************************************
+
+    Entry point for the agora-checkvtable
+        It generate C++ code for chacking the offset of the vtable.
+        The path to the file it creates is `source/scpp/extra/DVMChecks.cpp`.
+
+        The use of this is as follows.
+        agora-checkvtable source/scpp/extra/DVMChecks.cpp
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.cli.checkvtable.main;
+
+import scpd.tests.VTableTypes;
+import scpd.tests.VTableTest;
+
+import std.file;
+import std.format;
+import std.stdio;
+
+/// Workaround for issue likely related to dub #225,
+/// expects a main() function and invokes it after unittesting.
+version (unittest) void main () { } else:
+
+/*******************************************************************************
+
+    Application entry point
+
+*******************************************************************************/
+
+int main (string[] args)
```

Comment:
The entire body of this function should be another function which either returns a `string` or writes to a `buffer`. That way, you can unittest the function and see that it generates the right code.

Reply:
And then in `main()` you would just write that buffer to a file.

Reply:
@AndrejMitrovic That's a good idea. Thank you.

Reply:
I tried this.
There is an error message that the computer time cannot be accessed. 
I could not change to another string array instead of VTableCheckClasses.

Reply:
> There is an error message that the computer time cannot be accessed.

Could you show the full error message and also the code which is causing the issue?

Reply:
I fixed the source as below, and when I tested it, an error occurred.

``` D
/// This is agora-checkvtable sub-function for generating C++ source code
int generate (string[] args, ref string[] outputs)
{
    if (args.length < 3)
    {
        writeln("usage: agora-gen command [output source file]");
        return 1;
    }

    auto filename = args[2];
    writecode(VTableCheckClasses, outputs);
    File file = File(filename, "w+");

    foreach(ref line; outputs)
        file.writeln(line);

    file.close();

    writefln("Source code was generated with the file name %s.", filename);
    return 0;
}

/// Writes C++ source code to outputs
void writecode (immutable(string[]) types, ref string[] outputs)
{
    writeHeader(outputs);

    static foreach (e; types)
    {
        mixin(
        q{
            writeMethodHeader(outputs, `%1$s`, `getVMOffset%1$s`);
            static foreach (idx, member; __traits(allMembers, %1$s))
            {
                mixin(
                q{
                    static if (__traits(isVirtualMethod, %1$s.%2$s) && (`%2$s` != `__dtor`) && (`%2$s` != `__xdtor`))
                    {
                        writeMethodContents(outputs, `%1$s`, `%2$s`);
                    }
                }.format(`%1$s`, member));
            }

            writeMethodFooter(outputs);
            writeCheckFunction(outputs, `%1$s`);
            outputs ~= ``;

        }.format(e, `%2$s`));
    }

    outputs ~= `int checkVMOffset (const char* classname, const char* offsets)`;
    outputs ~= `{`;

    static foreach (e; types)
    {
        mixin(
        q{
            outputs ~= `    if (strcmp(classname, "%1$s") == 0)`;
            outputs ~= `    {`;
            outputs ~= `        return checkVMOffset%1$s(offsets);`;
            outputs ~= `    }`;
        }.format(e));
    }

    outputs ~= `    return -1;`;
    outputs ~= `}`;
}
```

source/agora/cli/checkvtable/generate.d(55,24): Error: variable types cannot be read at compile time
source/agora/cli/checkvtable/generate.d(81,24): Error: variable types cannot be read at compile time

Reply:
I solved this by using CTFE.